src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -196,8 +199,27 @@ private static MethodSymbol GetEntryPoint(CSharpCompilation compilation, PEModul
             diagnostics.AddRange(entryPointAndDiagnostics.Diagnostics);
 
             var entryPoint = entryPointAndDiagnostics.MethodSymbol;
-            var synthesizedEntryPoint = entryPoint as SynthesizedEntryPointSymbol;
-            if (((object)synthesizedEntryPoint != null) &&
+
+            if (entryPoint == null)
+            {
+                Debug.Assert(entryPointAndDiagnostics.Diagnostics.HasAnyErrors() || !compilation.Options.Errors.IsDefaultOrEmpty);
+                return null;
+            }
+
+            SynthesizedEntryPointSymbol synthesizedEntryPoint = null;
+            bool addedDefinition = false;
+            if (entryPoint is SynthesizedEntryPointSymbol s)
```

Comment:
`is SynthesizedEntryPointSymbol s` followed by `synthesizedEntryPoint = s;` seems more convoluted (to me at least) than:
```
var synthesizedEntryPoint = entryPoint as SynthesizedEntryPointSymbol;
if ((object)synthesizedEntryPoint == null)
{
    // other cases
}
```